includes/Checker/Checks/Plugin_Updater_Check.php

```diff
@@ -0,0 +1,187 @@
+
+		// Looks for "UpdateURI" in plugin header.
+		$this->look_for_update_uri_header( $result );
+
+		// Looks for special updater file.
+		$this->look_for_updater_file( $result, $php_files );
+
+		// Looks for plugin updater code in plugin files.
+		$this->look_for_plugin_updaters( $result, $php_files );
+
+		// Looks for plugin updater routines in plugin files.
+		$this->look_for_updater_routines( $result, $php_files );
+	}
+
+	/**
+	 * Looks for UpdateURI in plugin header and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 */
+	protected function look_for_update_uri_header( Check_Result $result ) {
+		if ( ! function_exists( 'get_plugin_data' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugin_main_file = $result->plugin()->abspath();
+		$plugin_header    = get_plugin_data( $plugin_main_file );
+		if ( ! empty( $plugin_header['UpdateURI'] ) ) {
+			$this->add_result_error_for_file(
+				$result,
+				$plugin_main_file,
+				__( 'Plugin Updater detected. Use of the Update URI header is not helpful in plugins hosted on WordPress.org.', 'plugin-check' )
+			);
+		}
+	}
+
+	/**
+	 * Looks for plugin updater file and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result    The check result to amend, including the plugin context to check.
+	 * @param array        $php_files List of absolute PHP file paths.
+	 */
+	protected function look_for_updater_file( Check_Result $result, array $php_files ) {
+
+		$filenames     = array_map( 'strtolower', array_map( 'basename', $php_files ) );
+		$blocked_files = array( 'plugin-update-checker.php' );
+		$matches       = array_intersect( $filenames, $blocked_files );
+
+		if ( $matches ) {
+			$this->add_result_error_for_file(
+				$result,
+				'plugin-update-checker.php',
+				sprintf(
+					/* translators: %s: The match updater file name. */
+					__( 'Plugin Updater detected. These are not permitted in WordPress.org hosted plugins. Detected: %s', 'plugin-check' ),
+					esc_html( implode( ', ', $matches ) )
+				)
+			);
+		}
+	}
+
+	/**
+	 * Looks for plugin updater code in plugin files and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result    The check result to amend, including the plugin context to check.
+	 * @param array        $php_files List of absolute PHP file paths.
+	 */
+	protected function look_for_plugin_updaters( Check_Result $result, array $php_files ) {
+
+		$needles = array(
+			"'plugin-update-checker'",
+			'WP_GitHub_Updater',
+			'WPGitHubUpdater',
+			'#class [A-Z_]+_Plugin_Updater#i',
+			'#updater\.\w+\.\w{2,5}#i',
+			'site_transient_update_plugins',
+		);
+
+		foreach ( $needles as $needle ) {
+			$match = self::file_str_contains( $php_files, $needle );
```

Comment:
Some of the strings in `$needles` are regular expressions, so for those `file_str_contains()` doesn't work. Instead you need to use `file_preg_match()` for them.

Reply:
Nice catch. Missed it 😞